internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -170,6 +176,40 @@ func (r *AgentServiceConfigReconciler) Reconcile(origCtx context.Context, req ct
 		return reconcile.Result{}, nil
 	}
 
+	if instance.DeletionTimestamp.IsZero() {
+		if !funk.ContainsString(instance.GetFinalizers(), agentServiceConfigFinalizerName) {
+			controllerutil.AddFinalizer(instance, agentServiceConfigFinalizerName)
+			if err := r.Update(ctx, instance); err != nil {
+				log.WithError(err).Error("failed to add finalizer to AgentServiceConfig")
+				return ctrl.Result{Requeue: true}, err
+			}
+		}
+	} else {
+		// do cleanup and remove finalizer
+		statefulSet := &appsv1.StatefulSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      imageServiceName,
+				Namespace: r.Namespace,
+			},
+		}
+		if err := r.Get(ctx, client.ObjectKeyFromObject(statefulSet), statefulSet); err != nil && !errors.IsNotFound(err) {
+			log.WithError(err).Error("failed to get image service stateful set for cleanup")
```

Comment:
If it no longer exists for whatever reason, what's the expectation?

I think once the StatefulSet is gone, we can't reliably determine if PVs were cleaned up or not, and should just remove the finalizer from ACI.

Reply:
Also I see below that if the statefulset is being deleted, that code path also calls `cleanupImageServiceFinalizer`. It's probably better to call that from just one place, unless there's a reason it needs to be in both?

Reply:
> If it no longer exists for whatever reason, what's the expectation?

I suppose this could be handled better. I do have a condition there to ensure that we don't return if the error returned by the `Get` call is `NotFound`, but it's probably best to not assume that `statefulSet` is in any way valid after that call.
I could also wrap the `cleanupImageServiceFinalizer` call in the same condition.

> Also I see below that if the statefulset is being deleted, that code path also calls cleanupImageServiceFinalizer. It's probably better to call that from just one place, unless there's a reason it needs to be in both?

I think it's better to have it in both places. Even if we force a requeue here after removing the finalizer on ASC then we could end up in a situation where we try to run the reconcile loop without an ASC instance and I think we fail on that rather early in this function. I don't think I'd want to let that go through just to not have to call this function again.

Reply:
Ah I missed the `!errors.IsNotFound(err)` part. Ok. But if it doesn't exist, does it make sense to still call `cleanupImageServiceFinalizer`?

Reply:
> But if it doesn't exist, does it make sense to still call cleanupImageServiceFinalizer?

This is what I was trying to address in the first part of my last comment.
If we assume that a failed `Get` call doesn't change `statefulSet` then it's safe to call that function because the finalizer won't be present on the little stub object and we'll just return. But I can make that explicit here by re-checking `errors.IsNotFound(err)` before calling the cleanup.